sdk/src/main/scala/com/horizen/SidechainMemoryPool.scala

```diff
@@ -109,12 +132,37 @@ class SidechainMemoryPool(unconfirmed: TrieMap[String, SidechainMemoryPoolEntry]
     }
 
     for (t <- txs) {
-      unconfirmed.put(t.id, SidechainMemoryPoolEntry(t))
+      val entry = SidechainMemoryPoolEntry(t)
+      if (entry.feeRate.getFeeRate() >= minFeeRate) {
+        addWithSizeCheck(entry)
+      }
     }
-
     new Success[SidechainMemoryPool](this)
   }
 
+  /**
+   * Add tx to the mempool.
+   * Lowest fee-rate txs are removed in case max mempool size is reached.
+   * @return true if the transaction has been inserted succesfully
+   */
+  def addWithSizeCheck(entry: SidechainMemoryPoolEntry) : Boolean = {
+    val orderedEntries = unconfirmed.values.toSeq.sortWith((a,b) => {
+      a.feeRate.getFeeRate() < b.feeRate.getFeeRate()
+    })
+    while (usedPoolSizeBytes + entry.feeRate.getSize() > maxPoolSizeBytes){
+      val lastEntry = orderedEntries.take(1).last
```

Comment:
`orderedEntries` are ordered from lowest fee rate to highest one, right?

`orderedEntries.take(1).last` can be replaced with `orderedEntries.head`

Moreover, in case mempool was empty and the very first tx exceed the limit (for example, it was set to 0 Mb), we will get an exception here.
Use `headOption` and check if entry present.